# Lec 4: More Integration

12 April 2018  

Puzzle from last time:

In [1]:
x = 1E15 # one with fifteen zeros
y = 1000000000000001.2345678 # x + 1.2345678

In [2]:
print(y-x) # should be 1.2345678

1.25


**Puzzle**: why is this 1.25? Sure, there's some rounding error. But shouldn't it round to either 1.23 (round down) or 1.24 (round up)? 

**Answer**: see hand-written lecture notes.

## Integral Algorithms

*General idea:* we want the area under a curve. We have a function, which we view as a tool for giving us **samples** of the curve. Given an $x_i$, the function spits out $f(x_i)$. Using this tool, we wrote up a code that:
1. Has some kind of loop to sample the function at discrete $x$ values
2. Sums together the approximate areas of under the curve from $x_i$ to $x_{i+1}$.

A reminder:

In [3]:
# Let's write a function that takes N samples in the range x in [a,b].
# (This way we can control how long it takes)

lower_limit = 1.0
upper_limit = 2.3

def integrate_Nsteps(function_name, num_steps):
    """Riemann sum from lower_limit to upper_limit"""
    delta_x = (upper_limit - lower_limit)/num_steps
    sample_at = lower_limit + (delta_x/2.)
    total = 0.
    
    while sample_at < upper_limit:
        total += function_name(sample_at)*delta_x
        sample_at = sample_at + delta_x
    
    return total

In [4]:
def myFun(x):
    return 3*x**2

integrate_Nsteps(myFun,100)
# Should be 11.167
# How did I know? I used Mathematica.

11.166945074999926

In [5]:
11.167 - _ # underscore means "previous output"

5.492500007342471e-05

Something to think about: how many times did we go through the `while` loop?
* $N-1$
* $N$
* $N+1$

(Correct answer: $N$. Try it with $N=2$)

Let's now try this with the trapezoidal rule.

In [6]:
# We can start by copy and pasting...

# Copy this over so that we don't have to depend on a
# global variable from somewhere far away
lower_limit = 1.0
upper_limit = 2.3

def trapezoid_Nsteps(function_name, num_steps):
    """Trapezoidal rule from lower_limit to upper_limit"""
    delta_x = (upper_limit - lower_limit)/num_steps
    x_low = lower_limit # lower edge of trapezoid
    total = 0.
    
    while x_low < upper_limit:
        total += (function_name(x_low)+function_name(x_low+delta_x))*delta_x/2.
        x_low = x_low + delta_x
    
    return total

In [7]:
trapezoid_Nsteps(myFun,100)

11.374589245499925

In [8]:
11.167 - _

-0.20758924549992486

**What's going on?** It looks like the Riemann sum approximation is better! I believe this has to do with the convexity (curvature) of the function. Draw the function and convince yourself that the trapezoidal rule always over-estimates for this function!

## Simpson's rule

Simpson's rule approximates the function with parabolas for each segment. The parbola is defined by three samplings.

From the lecture 3 notes:

$$(\text{small area})_i = \frac{\Delta x}{3}
\left[
f(x_i - \Delta x)
+ 4 f(x_i)
+ f(x_i + \Delta x)
\right]
$$

In [9]:
# We can start by copy and pasting...

# Copy this over so that we don't have to depend on a
# global variable from somewhere far away
lower_limit = 1.0
upper_limit = 2.3

def Simpson_Nsteps(function_name, num_steps):
    """Simpson rule from lower_limit to upper_limit"""
    delta_x = (upper_limit - lower_limit)/num_steps
    x_low = lower_limit # lower edge of trapezoid
    total = 0.
    
    while x_low + 2*delta_x <= upper_limit: # see why we do this?
        sum_of_terms = function_name(x_low) \
            + 4*function_name(x_low + delta_x) \
            + function_name(x_low + 2*delta_x)
        total += delta_x / 3 * sum_of_terms
        x_low = x_low + 2*delta_x
    
    return total

In [10]:
Simpson_Nsteps(myFun, 100)

11.167000000000002

In [11]:
11.167-_

-1.7763568394002505e-15

Actually, this accuracy seems remarkable! *Is it?*

This is an accuracy on the order of $10^{-15}$, which is pretty close to the limits of our precision. 